sys/mips/82371AB.c

```diff
@@ -0,0 +1,148 @@
+/* 82371AB PCI-ISA bridge driver */
+
+#define KL_LOG KL_DEV
+#include <sys/mimiker.h>
+#include <sys/device.h>
+#include <sys/rman.h>
+#include <sys/bus.h>
+#include <mips/malta.h>
+#include <sys/pci.h>
+#include <sys/interrupt.h>
+#include <dev/piixreg.h>
+#include <dev/isareg.h>
+#include <sys/libkern.h>
+#include <sys/devclass.h>
+
+typedef struct intel_isa_state {
+  resource_t *io;
+  rman_t io_rman;
+} intel_isa_state_t;
+
+static resource_t *intel_isa_alloc_resource(device_t *dev, res_type_t type,
+                                            int rid, rman_addr_t start,
+                                            rman_addr_t end, size_t size,
+                                            rman_flags_t flags) {
+  assert(dev->bus == DEV_BUS_ISA);
+
+  rman_t *rman = NULL;
+  intel_isa_state_t *isa = dev->parent->state;
+  bus_space_handle_t bh = 0;
+
+  if (type == RT_IOPORTS) {
+    assert(start >= IO_ICUSIZE);
+    rman = &isa->io_rman;
+    bh = isa->io->r_bus_handle;
+  } else {
+    /* We don't know how to allocate this resource, but maybe someone above us
+     * does. In current context this is necessary to allocate interrupts,
+     * which aren't managed by this driver. */
+    return bus_alloc_resource(dev->parent, type, rid, start, end, size, flags);
+  }
+
+  resource_t *r =
+    rman_reserve_resource(rman, type, rid, start, end, size, 0, flags);
+
+  if (r == NULL)
+    return NULL;
+
+  /* We shouldn't do that for IRQs, but rn we'll never end up here when
+   * allocating a resouce */
+  r->r_bus_tag = generic_bus_space;
+  r->r_bus_handle = bh + resource_start(r);
+
+  if (flags & RF_ACTIVE) {
+    if (bus_activate_resource(dev, r)) {
+      resource_release(r);
+      return NULL;
+    }
+  }
+
+  return r;
+}
+
+static int intel_isa_activate_resource(device_t *dev, resource_t *r) {
+  if (r->r_type != RT_IOPORTS)
+    return bus_activate_resource(dev->parent, r);
```

Comment:
Please don't do that. The problem is that `bus_activate_resource` is a non trivial wrapper for a real activation function. Because of this delegation the wrapper will be called twice and thereby the `resource_activate` function will be called twice which bumps up an internal `range` activation counter twice. I know that you do the same for the deactivation function and hence it should work properly, but it's a dirty and confusing solution. I'm afraid you have to do the following
```C
  device_t *isab = dev->parent;
  return BUS_METHODS(isab->parent).activate_resource(isab, r);
```

Reply:
Of course you can just remove (thereby don't implement) the function (the same for deactivation).

Reply:
Of course you can just remove (thereby don't implement) the function (the same for deactivation) and handle the "ISA" case in `gt_pci_activate_resource` (`gt_pci_deactivate_resource`).

Reply:
I decided to not remove these functions as I don't want to rely on behaviour of other drivers too much, but I left them empty which should cover all the scenarios correctly.